Guide on Continuous Testing (#144)
NASA-AMMOS/slim: docs/guides/software-lifecycle/continuous-testing/README.md

---
toc_min_heading_level: 2
toc_max_heading_level: 5
---

# Continuous Testing

<pre align="center">A comprehensive guide to developing a continuous testing plan, implementation, and automation approach for your project using AI tools.</pre>

> 

Reviewer: Please add a screenshot here depicting the meaning or essence of your guide. 

See example [here](https://github.com/NASA-AMMOS/slim/blob/main/docs/guides/software-lifecycle/security/secrets-detection/README.md?plain=1#L5-L7).

- Providing frequent test results and feedback for development teams
- Facilitating quicker releases
- Fostering a culture of continuous improvement in software development

---
## Quick Start

Although there are many steps in setting up a continuous testing architecture, the most important step is to document a shared plan of approach for your team. 
- **If you already have continuous testing enabled for your project**: we recommend documenting your plan for internal and external team members' benefit within a `TESTING.md`. This will help you communicate your testing architecture so that team members can contribute easily to improve code through additional tests and automation. Copy/paste our below template and fill it in with your testing setup. Further steps are detailed in the (#step-by-step) guide section, which we encourage you to follow to make test writing and automation more efficient.  
- **If you are brand new to continuous testing**: we recommend starting with our (#step-by-step) guide to set up an efficient continuous testing plan for your project. 

**[⬇️ Download our TESTING.md Template](TESTING)** (see [example](TESTING-example) of template in action)

> 

Reviewer: Is it possible to have an extremely short, concise version of other aspects of our guide here? For example:
- Running LLM's to generate unit tests
- Recommending system test frameworks like Robot framework, and the suggested LLM approach 
- Tying it all together with a continuous integration pipeline

        DeployRelease[Deploy Software Release]
        IntegrationTests[Test Against System Tests]
    end

    Code --> UnitTests
    UnitTests -->|Yes| BuildComponents
    UnitTests -->|No| Code
    BuildComponents --> PublishArtifacts
    SingleIntegratedBuild -->|Yes| BuildIntegratedSoftware
    PublishArtifacts --> BuildIntegratedSoftware
    BuildIntegratedSoftware --> PublishIntegratedArtifact
    PullArtifacts --> DeployRelease
    DeployRelease --> IntegrationTests
    PublishArtifacts --> TestDeployment
    PublishIntegratedArtifact --> TestDeployment
```

##### Types of System Tests

You should outline the types of system tests you plan to implement in your `TESTING.md` file. We suggest the following types of tests to include:
- Testing for integration
  - Interaction between software components
  - Interaction with external services, provided files, exchange of messages, etc.
- Testing for security
  - See existing [SLIM security best practices](/slim/docs/category/security) as part of your software development workflow.
- Testing for performance and load
- Testing user interfaces for gaps and compliance against policies

##### Integration Tests Automation

> We recommend the following steps for integration tests automation:

Reviewer: This is good @yunks128 but I'd suggest enhancing it to be more directly tied to the above diagram, which basically implies CI. I may even color code that diagram to the CT vs CI parts.

1. Integrate your integration tests into a [Continuous Integration (CI) pipeline](/docs/guides/software-lifecycle/continuous-integration), allowing for automatic execution of tests upon code changes.
2. Ensure that each integration test can run independently without relying on the state or output of other tests. 

##### Security Tests Automation

To aid in security testing automation, we recommend two steps:
1. Add security testing to your developers' local coding environment via pre-commit (see [Unit Test Automation](#31-unit-test-automation) section above)
2. Enable existing [SLIM security best practices](/slim/docs/category/security) as part of your software development workflow.

> ##### Performance Tests Automation

Reviewer: Can we point to actual recommended tools or strategies described on the web? Also - please consider pointing to the testing-frameworks.md document

We recommend the following steps for performance test automation:
1. Integrate performance tests into your Continuous Integration and Continuous Delivery (CI/CD) pipeline to enable regular and automated execution of performance tests as part of the software delivery process.
2. Implement assertions or checks that flag any deviations from the baseline performance that exceed a predefined threshold. These assertions can be based on various metrics, such as response time deviations, throughput degradation, or resource utilization spikes.

> ##### User Interface Tests Automation

Reviewer: Same here - please consider pointing to the testing-frameworks.md document

We recommend the following steps for user interface test automation: 
1. Set up a CI pipeline to automatically trigger UI test execution upon code changes or at scheduled intervals.
2. Implement parallel test execution to automatically run multiple tests simultaneously and reduce overall execution time.

### 4. Maintain Your Tests

> Here are some best practices for maintaining your tests:

Reviewer: Would suggest revising this section to make it more actionable:
- Describe the triggers that cause someone to maintain / improve tests (e.g. code changes, new developers that need to be educated about the TESTING.md, new tools in the mix, new security vulnerabilities, etc.)
- Describe the artifacts in this guide that need to be maintained (e.g. `TESTING.md`, unit tests, system tests, automation system)

- Automating as many of your tests as possible to run automatically
- Providing frequent test results and feedback for development teams
- Facilitating quicker releases
- Fostering a culture of continuous improvement in software development

---
## Quick Start

Although there are many steps in setting up a continuous testing architecture, the most important step is to document a shared plan of approach for your team. 
- **If you already have continuous testing enabled for your project**: we recommend documenting your plan for internal and external team members' benefit within a `TESTING.md`. This will help you communicate your testing architecture so that team members can contribute easily to improve code through additional tests and automation. Copy/paste our below template and fill it in with your testing setup. Further steps are detailed in the (#step-by-step) guide section, which we encourage you to follow to make test writing and automation more efficient.  
- **If you are brand new to continuous testing**: we recommend starting with our (#step-by-step) guide to set up an efficient continuous testing plan for your project. 

> 

Reviewer: ```suggestion
Key Concepts to Get Started with:

```

The artifacts below this need a bit of context I think.

Here's our recommended approach to deciding the right model for you to use (see our full list of [recommended code generation models](continuous-testing/testing-frameworks/#for-test-code-generation)):

Is your code open source and permissively licensed?
- Yes: We recommend high-performance cloud-based LLM models (see our [recommended cloud-based models](testing-frameworks/#cloud-based-models))
- No: We recommend using locally-run LLM models (e.g., [codellama](https://ollama.com/library/codellama)) for sensitive data, such as those provided by the [Ollama](https://ollama.com/) tool, to protect your code from dissemination. (⚠️ These models may not perform as well as cloud-based bigger models, such as `GPT-4`. We recommend asking the LLM to generate a simple test template for you to fill out.) 

> #### Use Case 1: Writing Unit Tests using Downloadable, Open Source Code Generation Models

Reviewer: ```suggestion
#### 2.1 Example: Writing Unit Tests using Downloadable, Open Source Code Generation Models
```

Use similar formatting and language to section 1 examples.

    - **List/Array Operations:**
      "Create tests for functions that operate on lists/arrays, covering scenarios like element addition, removal, and list comprehensions."
    - **Exception Handling:**
      "Generate unit tests to ensure proper exception handling in functions that may encounter errors. Include tests for both expected and unexpected exceptions."
    - **Example Generated Unit Test:**
    ```python
    % ollama run codellama "Generate unit tests for a function/method that performs basic arithmetic operations (addition, subtraction, multiplication, division)"
    
    import unittest
    
    class TestArithmeticOperations(unittest.TestCase):
        def test_addition(self):
            result = add(2, 3)
            self.assertEqual(result, 5)
    
        def test_subtraction(self):
            result = subtract(5, 3)
            self.assertEqual(result, 2)
    
        def test_multiplication(self):
            result = multiply(4, 6)
            self.assertEqual(result, 24)
    
        def test_division(self):
            result = divide(10, 5)
            self.assertEqual(result, 2)
    
    if __name__ == "__main__":
        unittest.main()
    ```

> #### Use Case 2: Writing System Tests using Downloadable, Open Source Code Generation Models and Behavior-Driven Development Frameworks

Reviewer: ```suggestion
#### 2.2 Example: Writing System Tests using Downloadable, Open Source Code Generation Models and Behavior-Driven Development Frameworks
```

    - **Results Location:** deployment test server's `/var/log/myapp/performance-tests.log
- **Framework Used:** Apache JMeter.
- **Tips:** Simulate peak load scenarios, monitor resource usage, and use tools like Chaos Monkey for resilience testing.

### 2. Write Your Tests

We recommend using [Testing Frameworks](testing-frameworks#for-code-development) when writing your tests to automate, organize, and analyze your testing efforts effectively. You can write your test code from scratch, but another way to write it is using large language models (LLMs), and specifically our recommended [Test Code Generation Frameworks](testing-frameworks#for-test-code-generation). 

Recent studies show that large language models (LLMs) can generate test code covering up to 85% of scenarios (source: [study](https://arxiv.org/pdf/2305.00418.pdf)). Open-source LLM tools like [codellama](https://ollama.com/library/codellama) can create initial test code, which developers can refine. Running these models locally addresses data privacy concerns. For the latest advancements, refer to the [code model ranking](https://huggingface.co/spaces/bigcode/bigcode-models-leaderboard).

> Here's our recommended approach to deciding the right model for you to use (see our full list of [recommended code generation models](continuous-testing/testing-frameworks/#for-test-code-generation)):

Reviewer: ```suggestion
Here's our recommended approach to deciding the right model for you to use (see our full list of [recommended code generation models](testing-frameworks#for-test-code-generation)):
```

@yunks128 - this is the broken link. The above fix should resolve it.

  - **Automatically:**
    - **Frequency:** Upon every commit or pull request.
    - **Results Location:** [GitHub Actions Unit Tests Workflow](#)
- **Framework Used:** Jest for JavaScript.
- **Tips:** Focus on core functions and methods, use mocks for dependencies, and handle edge cases.

#### 1.3 Example: Performance Testing

In the "Performance Testing" section, detail how you ensure your application can handle expected and peak loads. For example:

- **Location:** `/tests/performance`
- **Purpose:** Validate that the application meets performance goals under load.
- **Running Tests:**
  - **Manually:**
    1. Navigate to the `/tests/performance` directory.
    2. Run the performance testing tool command (e.g., `jmeter -n -t test.jmx`).
    3. Review generated reports for analysis.
  - **Automatically:**
    - **Frequency:** Monthly stress tests or before release candidates.
    - **Results Location:** deployment test server's `/var/log/myapp/performance-tests.log
- **Framework Used:** Apache JMeter.
- **Tips:** Simulate peak load scenarios, monitor resource usage, and use tools like Chaos Monkey for resilience testing.

### 2. Write Your Tests

> We recommend using [Testing Frameworks](testing-frameworks#for-test-code-generation) when writing your tests to automate, organize, and analyze your testing efforts effectively. You can write your test code from scratch, but another way to write it is using large language models (LLMs). 

Reviewer: ```suggestion
We recommend using [Testing Frameworks](testing-frameworks#for-code-development) when writing your tests to automate, organize, and analyze your testing efforts effectively. You can write your test code from scratch, but another way to write it is using large language models (LLMs), and specifically our recommended [Test Code Generation Frameworks](testing-frameworks#for-test-code-generation). 
```

Recent studies show that large language models (LLMs) can generate test code covering up to 85% of scenarios (source: [study](https://arxiv.org/pdf/2305.00418.pdf)). Open-source LLM tools like [codellama](https://ollama.com/library/codellama) can create initial test code, which developers can refine. Running these models locally addresses data privacy concerns. For the latest advancements, refer to the [code model ranking](https://huggingface.co/spaces/bigcode/bigcode-models-leaderboard).

> Here's our recommended approach to deciding the right model for you to use:

Reviewer: ```suggestion
Here's our recommended approach to deciding the right model for you to use (see our full list of [recommended code generation models](continuous-testing/testing-frameworks/#for-test-code-generation)):
```

Is your code open source and permissively licensed?

> - Yes: We recommend high-performance cloud-based LLM models (e.g., [GPT-4](https://openai.com/gpt-4), [Github Copilot](https://github.com/features/copilot))

Reviewer: ```suggestion
- Yes: We recommend high-performance cloud-based LLM models (see our [recommended cloud-based models](testing-frameworks/#cloud-based-models))
```

> - No: We recommend using locally-run LLM models (e.g., [codellama](https://ollama.com/library/codellama)) for sensitive data, such as those provided by the [Ollama](https://ollama.com/) tool, to protect your code from dissemination. (⚠️ These models may not perform as well as cloud-based bigger models, such as `GPT-4`. We recommend asking the LLM to generate a simple test template for you to fill out.) 

Reviewer: ```suggestion
- No: We recommend using downloadable, open source LLM models (see our [recommended downloadable open source models](continuous-testing/testing-frameworks/#downloadable-open-source-models) to protect your code from dissemination. 
    ⚠️ These models may not perform as well as cloud-based bigger models, such as `GPT-4`. We recommend asking the LLM to generate a simple test template for you to fill out.) 
```

> #### Use Case 1: Writing Unit Tests using LLMs

Reviewer: ```suggestion
#### Use Case 1: Writing Unit Tests using Downloadable, Open Source Code Generation Models
```

    - **List/Array Operations:**
      "Create tests for functions that operate on lists/arrays, covering scenarios like element addition, removal, and list comprehensions."
    - **Exception Handling:**
      "Generate unit tests to ensure proper exception handling in functions that may encounter errors. Include tests for both expected and unexpected exceptions."
    - **Example Generated Unit Test:**
    ```python
    % ollama run codellama "Generate unit tests for a function/method that performs basic arithmetic operations (addition, subtraction, multiplication, division)"
    
    import unittest
    
    class TestArithmeticOperations(unittest.TestCase):
        def test_addition(self):
            result = add(2, 3)
            self.assertEqual(result, 5)
    
        def test_subtraction(self):
            result = subtract(5, 3)
            self.assertEqual(result, 2)
    
        def test_multiplication(self):
            result = multiply(4, 6)
            self.assertEqual(result, 24)
    
        def test_division(self):
            result = divide(10, 5)
            self.assertEqual(result, 2)
    
    if __name__ == "__main__":
        unittest.main()
    ```

> #### Use Case 2: Writing System Tests using LLMs and Behavior-Driven Development (Robot Framework)

Reviewer: ```suggestion
#### Use Case 2: Writing System Tests using Downloadable, Open Source Code Generation Models and Behavior-Driven Development Frameworks
```

> System-level tests that involve interaction with external software or files can be complicated to test. Please refer to our [Testing Frameworks guide](testing-frameworks#for-system-testing) for system testing tools. We recommend leveraging the behavior-driven development (BDD), such as [**Robot Framework**](https://github.com/robotframework/QuickStartGuide/blob/master/QuickStart.rst) in collaboration with the Large-Language Models to meet the challenge. Specifically, using LLMs to auto-generate Robot Framework pseudocode can help streamline the process of creating integration test cases. Here's an example:

Reviewer: ```suggestion
System-level tests that involve interaction with external software or files can be complicated to test. Please refer to our [System Testing section within our Testing Frameworks Guide](testing-frameworks#for-system-testing) for recommended system testing tools that leverage behavior-driven development (BDD) philosophies. One such example, [**Robot Framework**](https://github.com/robotframework/QuickStartGuide/blob/master/QuickStart.rst), is demoed below in collaboration with Large-Language Models. Specifically, using LLMs to auto-generate Robot Framework pseudocode can help streamline the process of creating integration test cases. Complicated system testing scenarios can be simplified and written automatically using LLMs. Here's an example:
```

Login with MFA                                                        | PASS |
------------------------------------------------------------------------------
Tests.Mfa Login                                                       | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  MFA_Login/output.xml
Log:     MFA_Login/log.html
Report:  MFA_Login/report.html
```

The quality of the generated test code can be compared with the example provided by Robot Framework [here](https://docs.robotframework.org/docs/examples/mfa_login). The website name that you are testing needs to be updated, but other than that, it was perfect! 

### 3. Automate Your Tests

Our recommendation is to automate as many of your tests as possible. For tests that can't be automated, we suggest scheduling specific times for personnel to run manual tests.

#### 3.1 Static Test Automation

> We recommend setting up a static test using `.pre-commit-config.yaml`. A working example is [here](.pre-commit-config.yaml). 

Reviewer: We're describing the use of pre-commit tool but we haven't introduced it yet. This is a bit confusing. Note that the tool is introduced below in Section 3.2's Developer Machine section. I recommend rewriting Section 3.1 and Section 3.2 so that the tool is introduced before it is shown used.

Reviewer: @yunks128 - perhaps you could just add a quick note in the 3.1 section that further details about pre-commit will be discussed in the Developer Machine section.

##### User Interface (UI) Tests

In this section, you'll detail the UI testing process, including how these tests ensure compliance with user interaction needs. Mention the framework used for UI testing (e.g., Selenium). Provide information on the location of UI tests within the project repository and the testing schedule.

Include specifics about your testing setup in this template section as follows:

- `[INSERT PATH TO UI TEST FOLDER ON REVISION CONTROL]`: Direct where UI tests can be found in your project, for instance, /tests/system/ui.
- `[INSERT YOUR INTEGRATION TESTING FRAMEWORK OF CHOICE]`: Mention the framework or tool used for V&V testing, such as "TestRail for test management", and provide a link to how to use it within your project. Consult our [Testing Frameworks](testing-frameworks) guide for recommended frameworks to choose from.

### 2. Write Your Tests

> 

Reviewer: Writing tests requires a test framework - this would be a good place to reference your `testing-frameworks.md` to suggest the reader to select a testing framework appropriate to them and only after making that decision, to use an LLM to help generate tests.

Author: @riverma Thanks for your suggestion! I was about to write another paragraph in "2. Write Your Tests" section, and I found `testing-frameworks.md` was referenced many times in the prior section. Currently, we recommended `testing-frameworks.md` during `TESTING.md` test plan stage. What do you think?

Reviewer: @yunks128 - could you reference the snippet you're talking about? I think the idea is: the unit / system tests that could be written by an LLM need to run in a framework. We haven't explicitly mentioned that writing unit tests via an LLM needs to be augmented by an existing testing framework in which to run them. Its pretty obvious, yes, but the question could come up with novel technologies such as codelama. Basically: now that I've written a unit test, "now what"?

Writing comprehensive tests for large software applications can be a time-consuming process. Recent studies suggest that Large Language Models (LLMs) offer a strong solution to automatically generating test files, in some cases being able to cover up to 85% of test scenarios automatically (source: [study](https://arxiv.org/pdf/2305.00418.pdf)). We therefore strongly recommend using LLM automation to generate test cases, which are covered below. 

> We recommend leveraging open-source LLM-based tools like [codellama](https://ollama.com/library/codellama) to rapidly generate initial test code, which developers can then refine and expand as needed. These open-source models can run completely locally, so you don't need to worry about data privacy. The model landscape is evolving rapidly, so we suggest referring to the code model ranking to stay updated on the latest advancements. You can find the ranking at https://huggingface.co/spaces/bigcode/bigcode-models-leaderboard.

Reviewer: Also - many times the performance (as I've seen of codellama) will be insufficient and the tests simply won't be generated well especially with codellama - whereas GPT-3 might be better. I'd include a "⚠️ Warning" note in your section here that if you have a lot of trouble generating test code using codellama, ask the LLM to generate a simple test template (with no tests) for you to fill out. We should test that prompt on various code files to ensure it can universally generate a template well.

---
title: Continuous Testing
description: A comprehensive guide to continuous testing in software development.
sidebar_label: Continuous Testing
---

<!--
import CodeBlock from '@theme/CodeBlock';
import PreCommitConfigSource from '!!raw-loader!./.pre-commit-config.yaml';
-->

<pre align="center">A comprehensive guide to developing a continuous testing plan, implementation, and automation approach for your project using AI tools.</pre>

<div align="center">

>   <img src="/static/img/continuous-testing-image.png" alt="continuous-testing-representative-image"/>

Reviewer: Image not rendering in your example page.

  - **Automatically:**
    - **Frequency:** Upon every commit or pull request.
    - **Results Location:** [GitHub Actions Unit Tests Workflow](#)
- **Framework Used:** Jest for JavaScript.
- **Tips:** Focus on core functions and methods, use mocks for dependencies, and handle edge cases.

#### 1.3 Example: Performance Testing

In the "Performance Testing" section, detail how you ensure your application can handle expected and peak loads. For example:

- **Location:** `/tests/performance`
- **Purpose:** Validate that the application meets performance goals under load.
- **Running Tests:**
  - **Manually:**
    1. Navigate to the `/tests/performance` directory.
    2. Run the performance testing tool command (e.g., `jmeter -n -t test.jmx`).
    3. Review generated reports for analysis.
  - **Automatically:**
    - **Frequency:** Monthly stress tests or before release candidates.
    - **Results Location:** deployment test server's `/var/log/myapp/performance-tests.log
- **Framework Used:** Apache JMeter.
- **Tips:** Simulate peak load scenarios, monitor resource usage, and use tools like Chaos Monkey for resilience testing.

### 2. Write Your Tests

> We recommend using [Testing Frameworks](testing-frameworks) when writing your tests to automate, organize, and analyze your testing efforts effectively. You can write your test code from scratch, but another way to write it is using large language models (LLMs). 

Reviewer: We should link to the specific Code Generation Testing Frameworks section rather than the whole file. Right now it is a bit confusing which part of the Testing Frameworks file the reader should look at. You might want to add anchor links to that file, which would require reformatting to a Markdown page rather than the bullet point list format.

> We recommend adding inline comments in your tests to clarify the purpose of each test. These comments should include details on the function being tested, the test type (e.g., bug fix, change request, requirements validation, anomaly reports), and any relevant context. For example: 

Reviewer: The inline comments block is too long and placed in a bit of a confusing spot. You're talking about LLMs prior to that block and immediately after, so it might make sense to put that block elsewhere and shorten it.

Here's our recommended approach to deciding the right model for you to use:

Is your code open source and permissively licensed?
- Yes: We recommend high-performance cloud-based LLM models (e.g., [GPT-4](https://openai.com/gpt-4), [Github Copilot](https://github.com/features/copilot))
- No: We recommend using locally-run LLM models (e.g., [codellama](https://ollama.com/library/codellama)) for sensitive data, such as those provided by the [Ollama](https://ollama.com/) tool, to protect your code from dissemination. (⚠️ These models may not perform as well as cloud-based bigger models, such as `GPT-4`. We recommend asking the LLM to generate a simple test template for you to fill out.) 

#### Use Case 1: Writing Unit Tests using LLMs

>   - For unit tests, please follow the steps below: 

Reviewer: Formatting suggestions: we don't need bullet points here. Please see other instances of this in the README.md file.

    - **List/Array Operations:**
      "Create tests for functions that operate on lists/arrays, covering scenarios like element addition, removal, and list comprehensions."
    - **Exception Handling:**
      "Generate unit tests to ensure proper exception handling in functions that may encounter errors. Include tests for both expected and unexpected exceptions."
    - **Example Generated Unit Test:**
    ```python
    % ollama run codellama "Generate unit tests for a function/method that performs basic arithmetic operations (addition, subtraction, multiplication, division)"
    
    import unittest
    
    class TestArithmeticOperations(unittest.TestCase):
        def test_addition(self):
            result = add(2, 3)
            self.assertEqual(result, 5)
    
        def test_subtraction(self):
            result = subtract(5, 3)
            self.assertEqual(result, 2)
    
        def test_multiplication(self):
            result = multiply(4, 6)
            self.assertEqual(result, 24)
    
        def test_division(self):
            result = divide(10, 5)
            self.assertEqual(result, 2)
    
    if __name__ == "__main__":
        unittest.main()
    ```

> #### Use Case 2: Writing System Tests using LLMs and Robot Framework

Reviewer: This section should be updated to refer to Behavioral Driven Development (BDD) frameworks rather than Robot Framework specifically. The reason is that we're trying to recommend people to use BDD frameworks for system tests rather than say everyone should use Robot framework specifically. I think you'll want to add some other BDD frameworks, such as the one Unity uses, to your testing-frameworks.md and in this section mention to the reader they should scan the testing-frameworks.md section to select an appropriate BDD tool. You can still talk and give examples of Robot, but it should be more of an "Example" than a prescription. Similar to how you've done the LLM section and choice of models to use.

- `[INSERT YOUR INTEGRATION TESTING FRAMEWORK OF CHOICE]`: Mention the framework or tool used for V&V testing, such as "TestRail for test management", and provide a link to how to use it within your project. Consult our [Testing Frameworks](testing-frameworks) guide for recommended frameworks to choose from.

### 2. Write Your Tests

Writing comprehensive tests for large software applications can be a time-consuming process. While automatic test code generation tools like Pynguin show promise, they may not yet meet practical needs. Recent studies suggest that Large Language Models (LLMs) offer a viable solution, covering up to 85% of test methods automatically (source: [study](https://arxiv.org/pdf/2305.00418.pdf)).

Our recommendation is to leverage LLM-based tools, such as llama2, to quickly generate initial test code, with developers refining and expanding as necessary. Here's how:

1. **Download and Install OLLAMA:**
   - [OLLAMA](https://github.com/ollama/ollama): A streamlined tool for running LLMs locally.

3. **Invoke LLM and Generate Test Code:**
   ```bash

>    ollama run llama2 "$(cat ~/app_pack_generator/docker.py)" write a unit test code

Author: @riverma I tried to be more specific by mentioning a function in the script. Please find the output from llama2 and codellama: 

```
% ollama run llama2 "$(cat dswx_s1_validator.py) from the code provided above, write unit test for the function get_burst_id. do not explain the code, but provide the unit test script"

Here is a possible unit test script for the `get_burst_id` function:
```
```
import unittest
from typing import Optional

class TestGetBurstId(unittest.TestCase):
    def test_get_burst_id(self):
        # Test with a valid granule ID
        granule_id = 'T123-456-78901S1A'
        burst_id = get_burst_id(granule_id)
        self.assertIsNotNone(burst_id)

        # Test with an invalid granule ID
        invalid_granule_id = 'Invalid Granule ID'
        burst_id = get_burst_id(invalid_granule_id)
        self.assertIsNone(burst_id)

    def test_get_burst_id_empty(self):
        # Test with an empty string
        granule_id = ''
        burst_id = get_burst_id(granule_id)
        self.assertIsNone(burst_id)
```
```
In this test script, we define a `TestGetBurstId` class that inherits from the 
`unittest.TestCase` class. We then define three methods: `test_get_burst_id`, 
`test_get_burst_id_empty`, and `test_get_burst_id`.

The `test_get_burst_id` method tests the function `get_burst_id` with a valid 
granule ID. We create a `granule_id` variable with a valid ID, pass it to the 
`get_burst_id` function, and check that the resulting `burst_id` is not None.

The `test_get_burst_id_empty` method tests the function `get_burst_id` with an 
empty string. We create an `granule_id` variable with an empty string, pass it 
to the `get_burst_id` function, and check that the resulting `burst_id` is 
None.

The `test_get_burst_id` method is similar, but tests the function with an 
invalid granule ID. We create a `granule_id` variable with an invalid ID, pass 
it to the `get_burst_id` function, and check that the resulting `burst_id` is 
None.

By running these unit tests, we can ensure that the `get_burst_id` function 
works correctly and returns the expected values in different scenarios.
```

Codellama example:
```
% ollama run codellama "$(cat dswx_s1_validator.py) from the code provided above, write unit test for the function get_burst_id. do not explain the code, but provide the unit test script"

[PYTHON]
import re

def get_burst_id(granule_id):
    pattern = r'_T(\d+)-(\d+)-([A-Z]+\d+)_\d+T\d+Z_\d+T\d+Z_S1A_\d+_v\d+\.\d+'
    match = re.search(pattern, granule_id)
    if (match):
        return match.group()
    else:
        return None
[/PYTHON]
[TESTS]
# Test data
test_data = [
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_C", 
"T1-16-JYCZ"),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_Q", None),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C", None),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_R", 
"T1-16-JYCZ"),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_O", None),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_N", None),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_M", None),
    ("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_L", None),
]

# Test get_burst_id function
def test_get_burst_id():
    for granule, burst in test_data:
        assert get_burst_id(granule) == burst
[/TESTS]
[EXAMPLE]
# Example usage
assert 
get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_N")get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_0194C_N") == "T1-16-JYCZ"
# Example usage with multiple bursts
assert 
get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_N")get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_0194C_N") == "T1-16-JYCZ"
# Example usage with multiple bursts
assert 
get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_019B4C_M")get_burst_id("S1A_IW_GRDH_1SDV_20150729T181324_20150729T181346_012658_0194C_M") == "T1-16-JYCZ"
[/EXAMPLE]

```

>    ollama run llama2 "$(cat ~/app_pack_generator/docker.py)" write a unit test code

Author: @yunks128 
1. write tips section with step by step process of using LLM and how to refine the prompts (when the prompt didn't work for more complex code base, you need more specific prompts)  
2. mention other models in the guide - ChatGPT?  (proprietary)  Gemma? -> include in tools and frameworks page

        - Add additional steps.
        - Include assertions for edge cases.
        - Incorporate custom keywords or libraries.

3. **Test Execution**:
    - Run the tests locally or integrate them into your CI pipeline.

By combining LLM's natural language capabilities with Robot Framework's structured format, you can efficiently create and adapt  test cases.

##### 2.2.1 Example - Robot Framework and LLM Synergy 

Generate a Robot Framework script for MFA login with the following steps:

1. Navigate to the login page.
2. Fill in the username and password.
3. Generate a TOTP (Time-Based One-Time Password).
4. Fill in the TOTP code field.
5. Click the "Sign in" button.
6. Verify that the login was successful.

Here's a simple prompt to generate the script:

"Generate a Robot Framework script to perform MFA (Multi-Factor Authentication) login. The script should navigate to the login page, fill in the username and password fields, generate a TOTP code using the provided secret, enter the TOTP code, click the 'Sign in' button, and verify that the login was successful by checking the welcome message."

Here's the LLM output Robot Framework script:

> ```robot

Reviewer: Looks like this could be compared to the example from the robot framework docs: i.e. [here](https://docs.robotframework.org/docs/examples/mfa_login)? Might want to spell out how well the LLM did in this example?
